File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/binary_tree_inorder_traversal/BinaryTreeInorderTraversal4.java
package main.java.com.jiang.leetcode100.binary_tree_inorder_traversal;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
import java.util.Queue;
import java.util.Scanner;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) {this.val = val;}
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution4 {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if(root == null) return result;
        Deque<TreeNode> stack = new ArrayDeque<>();
        Deque<Integer> status = new ArrayDeque<>();

        stack.push(root);
        status.push(0);

        while(!stack.isEmpty()) {
            TreeNode cur = stack.pop();
            int cur_statu = status.pop();
            if(cur_statu == 0) {
                if(cur.right != null) {
                    stack.push(cur.right);
                    status.push(0);
                }
                stack.push(cur);
                status.push(1);
                if(cur.left != null) {
                    stack.push(cur.left);
                    status.push(0);
                }
            } else {
                result.add(cur.val);
            }
        }
        return result;
    }
}

public class BinaryTreeInorderTraversal4 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] nums = new  int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) {
            nums[i] = Integer.parseInt(nums_str[i]);
        }

        TreeNode root = new TreeNode(nums[0], null, null);
        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);

        for(int i = 1; i < nums.length; i += 2) {
            TreeNode cur = queue.poll();
            if(nums[i] != -1) {
                cur.left = new TreeNode(nums[i], null, null);
                queue.add(cur.left);
            }
            if(nums[i + 1] != -1) {
                cur.right = new TreeNode(nums[i + 1], null, null);
                queue.add(cur.right);
            }
        }

        List<Integer> result = new Solution4().inorderTraversal(root);
        for(Integer num : result) System.out.print(num + " ");
    }
}

File: Leetcode100/leetcode100_python/BinaryTreeInorderTraversal4.py
from typing import List

class TreeNode:
    def __init__(self, val = 0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        result = []
        stack = [(root, 0)]
        while stack:
            cur, status = stack.pop()
            if not cur:
                continue
            if status == 0:
                stack.append((cur.right, 0))
                stack.append((cur, 1))
                stack.append((cur.left, 0))
            elif status == 1:
                result.append(cur.val)
        return result

if __name__ == "__main__":
    nums = [int(num) for num in input().split()]
    root = TreeNode(nums[0])
    queue = [root]
    for i in range(1, len(nums), 2):
        cur = queue.pop(0)
        if nums[i] != -1:
            cur.left = TreeNode(nums[i])
            queue.append(cur.left)
        if nums[i + 1] != -1:
            cur.right = TreeNode(nums[i + 1])
            queue.append(cur.right)
    solution = Solution()
    print(solution.inorderTraversal(root))